util: polymorphic primatives (#1690)
pluralsight/design-system: docs/src/content/patterns/polymorphic.mdx

components that support polymorphism using the common `styled-components` api
patterns.

> Official [styled-component polymorphic docs](https://styled-components.com/docs/api#as-polymorphic-prop)

```typescript noRender startExpanded
import styled from "styled-components";
import { PolymorphicButton } from 'some/other/psds/component/path'

const StyledButton = styled(PolymorphicButton)`
  color: red;
`;

render(
  <>
    <StyledButton />
    <StyledButton as="a" />
    <StyledButton as="span" />
  </>
)
```

### Usage with Emotion

The Design System polymorphic approach is also designed to work with `emotion`.
You should be able to compose and style any Design System components that
supports polymorphism using the common `emotion` api patterns.

> Official [emotion polymorphic docs](https://emotion.sh/docs/styled#as-prop)

```typescript noRender startExpanded
import emotion from "@emotion/styled";
import { PolymorphicButton } from 'some/other/psds/component/path'

/* NOTE: I've had some issues with the emotion typings to fully support our
 *       polymorphic `as` typings. Below is the approach I've found works the

>  *       best and avoides type errors - although it's not as clean or as helpful

Reviewer: ```suggestion
 *       best and avoids type errors - although it's not as clean or as helpful
```

>  *       with it's introspections as the styled-component examples. I'm not as

Reviewer: ```suggestion
 *       with its introspections as the styled-component examples. I'm not as
```

 *       experienced with emotion as i would like to be...so if anyone reading
 *       this has a better approach please inform the Design System team or open
 *       an issue on github.
 *
 *       - danethurber 03/25/2021
 */
const BaseStyledButton = emotion("button")`
  color: hotpink;
`;
const StyledButton = React.forwardRef<HTMLButtonElement, any>((props, ref) => (
  <BaseStyledButton as={PolymorphicButton} ref={ref} {...props} />
));

render(
  <>
    <StyledButton />
    <StyledButton as="a" />
    <StyledButton as="span" />
  </>
)
```

## Creating a polymorphic component

The `utils` package of the design system exports a genericized higher order
function that you can use to replace the native `React.forwardRef`. Using the

> `forwardRefWithAs` utility with provide you with greater type intropection of

Reviewer: ```suggestion
`forwardRefWithAs` utility will provide you with greater type introspection of
```

the `ref`, native props/attribute typings, and will include typings for the
`as` props.

```typescript noRender startExpanded
import { forwardRefWithAs } from '@pluralsight/ps-design-system-util'

const Button = forwardRefWithAs<Props, 'button'>((props, ref) => {
  const { as: Comp = 'button', ...rest } = props
  return <Comp ref={ref} {...rest} />
})
```

If you need a component that also support static properties (like for instance
you are using the compound components pattern) we provide a higher order

> function to create a typesafe components with typed static properties.

Reviewer: ```suggestion
function to create typesafe components with typed static properties.
```
